File: hawk-server/src/Core/MetadataStore.cs
using System.Collections;
using System.Text;
using Tomlyn;
using Tomlyn.Model;

namespace Hawk.Server.Core;

/// <summary>
/// 元数据存储：.hawk/metadata/&lt;hash&gt;.toml 的读写。.hawk/metadata/ 是唯一权威数据源，
/// 参与网盘同步；写入采用「临时文件 + rename」的原子写，避免网盘同步到写了一半的文件。
///
/// 本类内存中保留权威副本（含 path → hash 反查表）供流水线热路径查询；
/// 副本的注水来源优先级：SQLite 派生缓存（IndexDb，快路径）→ TOML 全量解析（回退+建缓存）。
/// 只有索引流水线写入本存储，因此内部状态用一把锁保护即可。
/// </summary>
public sealed class MetadataStore
{
    private readonly LibraryPaths _paths;
    private readonly IndexDb _db;
    private readonly ILogger<MetadataStore> _logger;
    private readonly object _gate = new();

    private readonly Dictionary<string, ItemMetadata> _byHash = new();
    private readonly Dictionary<string, string> _hashByPath = new(); // 库内路径 → hash

    public MetadataStore(LibraryPaths paths, IndexDb db, ILogger<MetadataStore> logger)
    {
        _paths = paths;
        _db = db;
        _logger = logger;

        List<(string Hash, ItemMetadata Meta, long SourceMtime)>? entries = null;
        if (_db.IsHydrated)
        {
            try
            {
                entries = _db.LoadAll();
                _logger.LogInformation("元数据副本已从 SQLite 缓存注水 {Count} 条", entries.Count);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "元数据缓存读取失败，改由 TOML 全量解析");
            }
        }

        if (entries is null)
        {
            // 缓存缺失/未注水/读取失败：TOML 全量解析（一次性慢路径），并顺带建好缓存
            entries = LoadAllFromToml();
            _db.Hydrate(entries);
        }

        HydrateMemory(entries);
    }

    public bool TryGet(string hash, out ItemMetadata meta)
    {
        lock (_gate)
        {
            return _byHash.TryGetValue(hash, out meta!);
        }
    }

    /// <summary>全部元数据条目快照（批量迁移用）</summary>
    public KeyValuePair<string, ItemMetadata>[] Snapshot()
    {
        lock (_gate)
        {
            return _byHash.ToArray();
        }
    }

    /// <summary>按库内路径反查所属内容哈希（元数据 paths 记录）。</summary>
    public string? FindHashByPath(string libraryPath)
    {
        lock (_gate)
        {
            return _hashByPath.GetValueOrDefault(libraryPath);
        }
    }

    /// <summary>各 TOML 源文件 mtime 快照（后台对账比对依据）；缓存不可用时返回 null（本轮对账跳过）</summary>
    public IReadOnlyDictionary<string, long>? SourceMtimes() => _db.LoadSourceMtimes();

    /// <summary>
    /// 保存元数据：先 TOML 原子写（权威层），成功后更新内存副本与 SQLite 缓存。
    /// 中途崩溃时缓存与内存自然朝 TOML 收敛（后台对账会补齐），故必须先写 TOML。
    /// </summary>
    public void Save(string hash, ItemMetadata meta)
    {
        var file = FilePath(hash);
        var tmp = file + ".tmp";
        File.WriteAllText(tmp, Serialize(meta));
        File.Move(tmp, file, overwrite: true);
        var sourceMtime = LibraryPaths.ToUnixMs(File.GetLastWriteTimeUtc(file));

        lock (_gate)
        {
            _byHash[hash] = meta;
            RebuildPathIndex(hash, meta);
        }

        _db.Save(hash, meta, sourceMtime);
    }

    public void Delete(string hash)
    {
        var file = FilePath(hash);
        if (File.Exists(file))
        {
            File.Delete(file);
        }

        lock (_gate)
        {
            if (_byHash.Remove(hash, out var meta))
            {
                foreach (var p in meta.Paths)
                {
                    // 路径可能已迁移到其他 hash，只删仍指向本 hash 的映射
                    if (_hashByPath.TryGetValue(p.Path, out var owner) && owner == hash)
                    {
                        _hashByPath.Remove(p.Path);
                    }
                }
            }
        }

        _db.Delete(hash);
    }

    /// <summary>
    /// 对账应用（只进不出）：TOML 被外部新增/修改（网盘同步落地、手工编辑）后载入。
    /// 解析失败返回 false（跳过该文件，不清空任何状态，下一轮对账重试）。
    /// </summary>
    public bool ApplyExternalToml(string hash, string file, long sourceMtime)
    {
        ItemMetadata meta;
        try
        {
            meta = Parse(File.ReadAllText(file));
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "元数据解析失败，对账跳过: {File}", file);
            return false;
        }

        lock (_gate)
        {
            _byHash[hash] = meta;
            RebuildPathIndex(hash, meta);
        }

        _db.Save(hash, meta, sourceMtime);
        return true;
    }

    /// <summary>
    /// 对账应用（只进不出）：TOML 已消失 → 清空素材参数字段（等价于重启后无元数据的语义；
    /// item 本身与位置的存续由文件扫描决定，路径记录随之清空，扫描会按需重建空壳）。
    /// </summary>
    public void ClearExternal(string hash)
    {
        ItemMetadata meta;
        lock (_gate)
        {
            if (!_byHash.TryGetValue(hash, out meta!))
            {
                return;
            }

            meta.Url = null;
            meta.Star = 0;
            meta.Annotation = null;
            meta.Tags.Clear();
            meta.Categories.Clear();
            meta.Paths.Clear();
        }

        _db.Save(hash, meta, 0);
    }

    private string FilePath(string hash) => Path.Combine(_paths.MetadataDir, hash + ".toml");

    private void HydrateMemory(List<(string Hash, ItemMetadata Meta, long SourceMtime)> entries)
    {
        lock (_gate)
        {
            foreach (var (hash, meta, _) in entries)
            {
                _byHash[hash] = meta;
                RebuildPathIndex(hash, meta);
            }
        }
    }

    private void RebuildPathIndex(string hash, ItemMetadata meta)
    {
        foreach (var (path, h) in _hashByPath.Where(kv => kv.Value == hash).ToList())
        {
            _hashByPath.Remove(path);
        }

        foreach (var p in meta.Paths)
        {
            _hashByPath[p.Path] = hash;
        }
    }

    /// <summary>TOML 全量解析（缓存缺失时的权威回退路径）。只识别 64 位小写 hex 命名的文件，同步冲突副本等一律忽略。</summary>
    private List<(string Hash, ItemMetadata Meta, long SourceMtime)> LoadAllFromToml()
    {
        var entries = new List<(string, ItemMetadata, long)>();
        if (!Directory.Exists(_paths.MetadataDir))
        {
            return entries;
        }

        foreach (var file in Directory.EnumerateFiles(_paths.MetadataDir, "*.toml"))
        {
            var hash = Path.GetFileNameWithoutExtension(file);
            if (hash.Length != 64 || hash.Any(c => !char.IsAsciiHexDigitLower(c)))
            {
                continue;
            }

            try
            {
                var meta = Parse(File.ReadAllText(file));
                var mtime = LibraryPaths.ToUnixMs(File.GetLastWriteTimeUtc(file));
                entries.Add((hash, meta, mtime));
            }
            catch (Exception ex)
            {
                _logger.LogWarning(ex, "元数据解析失败，已跳过: {File}", file);
            }
        }

        _logger.LogInformation("已从 TOML 全量解析 {Count} 条元数据", entries.Count);
        return entries;
    }

    private static ItemMetadata Parse(string toml)
    {
        var table = TomlSerializer.Deserialize<TomlTable>(toml) ?? new TomlTable();
        var meta = new ItemMetadata();

        if (table.TryGetValue("url", out var url) && url is string urlStr)
        {
            meta.Url = urlStr;
        }

        if (table.TryGetValue("star", out var star) && star is long starNum)
        {
            meta.Star = (int)starNum;
        }

        if (table.TryGetValue("annotation", out var ann) && ann is string annStr)
        {
            meta.Annotation = annStr;
        }

        if (table.TryGetValue("tags", out var tags) && tags is IEnumerable tagArr)
        {
            meta.Tags = tagArr.Cast<object?>().OfType<string>().ToList();
        }

        if (table.TryGetValue("categories", out var cats) && cats is IEnumerable catArr)
        {
            meta.Categories = catArr.Cast<object?>().OfType<string>().ToList();
        }

        if (table.TryGetValue("paths", out var paths) && paths is IEnumerable pathArr)
        {
            foreach (var entry in pathArr.Cast<object?>())
            {
                if (entry is not IDictionary<string, object?> e || e["path"] is not string p)
                {
                    continue;
                }

                meta.Paths.Add(new PathEntry
                {
                    Path = p,
                    Size = e.TryGetValue("size", out var s) ? Convert.ToInt64(s) : 0,
                    ModificationTime = e.TryGetValue("modification_time", out var m) ? Convert.ToInt64(m) : 0,
                });
            }
        }

        return meta;
    }

    /// <summary>
    /// 序列化为 TOML。schema 固定，手写序列化以精确控制输出格式
    /// （标量在前、[[paths]] 在后，缺省字段省略）。
    /// </summary>
    private static string Serialize(ItemMetadata meta)
    {
        var sb = new StringBuilder();

        if (meta.Url is not null)
        {
            sb.Append("url = ").AppendLine(TomlString(meta.Url));
        }

        if (meta.Tags.Count > 0)
        {
            sb.Append("tags = [").Append(string.Join(", ", meta.Tags.Select(TomlString))).AppendLine("]");
        }

        if (meta.Categories.Count > 0)
        {
            sb.Append("categories = [").Append(string.Join(", ", meta.Categories.Select(TomlString))).AppendLine("]");
        }

        if (meta.Star > 0)
        {
            sb.Append("star = ").AppendLine(meta.Star.ToString());
        }

        if (meta.Annotation is not null)
        {
            sb.Append("annotation = ").AppendLine(TomlString(meta.Annotation));
        }

        foreach (var p in meta.Paths)
        {
            sb.AppendLine();
            sb.AppendLine("[[paths]]");
            sb.Append("path = ").AppendLine(TomlString(p.Path));
            sb.Append("size = ").AppendLine(p.Size.ToString());
            sb.Append("modification_time = ").AppendLine(p.ModificationTime.ToString());
        }

        return sb.ToString();
    }

    private static string TomlString(string value)
    {
        var sb = new StringBuilder(value.Length + 2);
        sb.Append('"');
        foreach (var c in value)
        {
            sb.Append(c switch
            {
                '\\' => "\\\\",
                '"' => "\\\"",
                '\n' => "\\n",
                '\r' => "\\r",
                '\t' => "\\t",
                _ when char.IsControl(c) => $"\\u{(int)c:X4}",
                _ => c,
            });
        }

        sb.Append('"');
        return sb.ToString();
    }
}

File: hawk-server-rs/src/core/metadata_store.rs
//! 元数据存储：.hawk/metadata/<hash>.toml 的读写 + 内存权威副本（含 path → hash 反查表）。
//! 写入采用「临时文件 + rename」的原子写；写入顺序铁律：先 TOML 成功后再写缓存与内存副本，
//! 中途崩溃自然朝 TOML 收敛。副本注水来源：IndexDb 快路径 → TOML 全量解析回退（顺带建缓存）。
//! 与 C# MetadataStore 语义一致。

use crate::core::index_db::IndexDb;
use crate::core::metadata::{self, ItemMetadata};
use crate::core::paths::{file_mtime_ms, LibraryPaths};
use std::collections::HashMap;
use std::sync::Mutex;

pub struct MetadataStore {
    paths: LibraryPaths,
    db: std::sync::Arc<IndexDb>,
    inner: Mutex<MetadataInner>,
}

#[derive(Default)]
struct MetadataInner {
    by_hash: HashMap<String, ItemMetadata>,
    hash_by_path: HashMap<String, String>,
}

impl MetadataStore {
    /// 构造即注水：SQLite 缓存快路径 → TOML 全量解析回退（顺带建好缓存）
    pub fn new(paths: LibraryPaths, db: std::sync::Arc<IndexDb>) -> MetadataStore {
        let mut entries: Option<Vec<(String, ItemMetadata, i64)>> = None;
        if db.hydrated.load(std::sync::atomic::Ordering::SeqCst) {
            match db.load_all() {
                Ok(loaded) => {
                    tracing::info!("元数据副本已从 SQLite 缓存注水 {} 条", loaded.len());
                    entries = Some(loaded);
                }
                Err(e) => tracing::error!("元数据缓存读取失败，改由 TOML 全量解析: {e}"),
            }
        }

        let entries = match entries {
            Some(e) => e,
            None => {
                let from_toml = load_all_from_toml(&paths);
                db.hydrate(&from_toml);
                from_toml
            }
        };

        let mut inner = MetadataInner::default();
        for (hash, meta, _) in entries {
            inner.by_hash.insert(hash.clone(), meta.clone());
            for p in &meta.paths {
                inner.hash_by_path.insert(p.path.clone(), hash.clone());
            }
        }
        MetadataStore {
            paths,
            db,
            inner: Mutex::new(inner),
        }
    }

    pub fn try_get(&self, hash: &str) -> Option<ItemMetadata> {
        self.inner.lock().unwrap().by_hash.get(hash).cloned()
    }

    /// 全部元数据条目快照（批量迁移用）
    pub fn snapshot(&self) -> Vec<(String, ItemMetadata)> {
        self.inner.lock().unwrap().by_hash.iter().map(|(h, m)| (h.clone(), m.clone())).collect()
    }

    /// 按库内路径反查所属内容哈希（元数据 paths 记录）
    pub fn find_hash_by_path(&self, library_path: &str) -> Option<String> {
        self.inner.lock().unwrap().hash_by_path.get(library_path).cloned()
    }

    /// 全量文件夹快照（增量扫描的对比基准；缓存不可用时返回空表 = 首轮全量深入）
    pub fn load_folder_snapshots(&self) -> HashMap<String, (i64, i64)> {
        self.db.load_folder_snapshots()
    }

    /// 整体替换文件夹快照（每轮扫描一次；遍历不完整时不调用）
    pub fn replace_folder_snapshots(&self, snapshots: &HashMap<String, (i64, i64)>) {
        self.db.replace_folder_snapshots(snapshots);
    }

    /// 各 TOML 源文件 mtime 快照（后台对账比对依据）；缓存不可用时返回 None（本轮对账跳过）
    pub fn source_mtimes(&self) -> Option<HashMap<String, i64>> {
        self.db.load_source_mtimes()
    }

    /// 保存元数据：先 TOML 原子写（权威层），成功后更新内存副本与 SQLite 缓存。
    /// TOML 写失败返回 Err（调用方决定回传 API 或仅记录，与 C# 异常传播语义一致）
    pub fn save(&self, hash: &str, meta: &ItemMetadata) -> Result<(), String> {
        let file = self.file_path(hash);
        let tmp = format!("{file}.tmp");
        std::fs::write(&tmp, metadata::serialize(meta)).map_err(|e| format!("元数据 TOML 写入失败 {tmp}: {e}"))?;
        if let Err(e) = std::fs::rename(&tmp, &file) {
            let _ = std::fs::remove_file(&tmp);
            return Err(format!("元数据 TOML 替换失败 {file}: {e}"));
        }
        let source_mtime = file_mtime_ms(&file);

        {
            let mut inner = self.inner.lock().unwrap();
            inner.by_hash.insert(hash.to_string(), meta.clone());
            rebuild_path_index(&mut inner, hash, meta);
        }
        self.db.save(hash, meta, source_mtime);
        Ok(())
    }

    pub fn delete(&self, hash: &str) {
        let file = self.file_path(hash);
        if std::path::Path::new(&file).is_file() {
            let _ = std::fs::remove_file(&file);
        }
        {
            let mut inner = self.inner.lock().unwrap();
            if let Some(meta) = inner.by_hash.remove(hash) {
                for p in &meta.paths {
                    if inner.hash_by_path.get(&p.path).map(|h| h == hash).unwrap_or(false) {
                        inner.hash_by_path.remove(&p.path);
                    }
                }
            }
        }
        self.db.delete(hash);
    }

    /// 对账应用（只进不出）：TOML 被外部新增/修改后载入。
    /// 解析失败返回 false（跳过该文件，不清空任何状态，下一轮对账重试）
    pub fn apply_external_toml(&self, hash: &str, file: &str, source_mtime: i64) -> bool {
        let text = match std::fs::read_to_string(file) {
            Ok(t) => t,
            Err(e) => {
                tracing::warn!("元数据读取失败，对账跳过: {file}: {e}");
                return false;
            }
        };
        let meta = match metadata::parse(&text) {
            Ok(m) => m,
            Err(e) => {
                tracing::warn!("元数据解析失败，对账跳过: {file}: {e}");
                return false;
            }
        };
        {
            let mut inner = self.inner.lock().unwrap();
            inner.by_hash.insert(hash.to_string(), meta.clone());
            rebuild_path_index(&mut inner, hash, &meta);
        }
        self.db.save(hash, &meta, source_mtime);
        true
    }

    /// 对账应用（只进不出）：TOML 已消失 → 清空素材参数字段（等价于重启后无元数据的语义）
    pub fn clear_external(&self, hash: &str) {
        let meta = {
            let mut inner = self.inner.lock().unwrap();
            match inner.by_hash.get_mut(hash) {
                Some(m) => {
                    m.url = None;
                    m.star = 0;
                    m.annotation = None;
                    m.tags.clear();
                    m.categories.clear();
                    m.paths.clear();
                    m.clone()
                }
                None => return,
            }
        };
        self.db.save(hash, &meta, 0);
    }

    fn file_path(&self, hash: &str) -> String {
        format!("{}/{}.toml", self.paths.metadata_dir, hash)
    }
}

fn rebuild_path_index(inner: &mut MetadataInner, hash: &str, meta: &ItemMetadata) {
    inner.hash_by_path.retain(|_, h| h != hash);
    for p in &meta.paths {
        inner.hash_by_path.insert(p.path.clone(), hash.to_string());
    }
}

/// TOML 全量解析（缓存缺失时的权威回退路径）
fn load_all_from_toml(paths: &LibraryPaths) -> Vec<(String, ItemMetadata, i64)> {
    let mut entries = Vec::new();
    let dir = &paths.metadata_dir;
    let read_dir = match std::fs::read_dir(dir) {
        Ok(d) => d,
        Err(_) => return entries,
    };
    for entry in read_dir.flatten() {
        let file_name = entry.file_name().to_string_lossy().to_string();
        if !file_name.ends_with(".toml") {
            continue;
        }
        let hash = file_name.strip_suffix(".toml").unwrap();
        if !metadata::is_valid_hash_file_name(hash) {
            continue;
        }
        let file = entry.path().to_string_lossy().to_string();
        let text = match std::fs::read_to_string(&file) {
            Ok(t) => t,
            Err(_) => continue,
        };
        match metadata::parse(&text) {
            Ok(meta) => {
                let mtime = file_mtime_ms(&file);
                entries.push((hash.to_string(), meta, mtime));
            }
            Err(e) => tracing::warn!("元数据解析失败，已跳过: {file}: {e}"),
        }
    }
    tracing::info!("已从 TOML 全量解析 {} 条元数据", entries.len());
    entries
}
